### PR TITLE
Lineage Registration: dynamic spacial registration

### LINKED ISSUE #22: Lineage Registration: improve registration of all timepoints
# User Story

Users want to compare embryos that move and rotate a lot. For these embryos it is probably not sufficient to compute an affine transform during the first time point. 

We need to compute an affine transform for each comparison of cell division direction to better compensate the embryo movement. 

## Potential Solution

The affine transform is currently calculated based on the root cell as landmarks. We could similarly create landmarks from the root cells descendants average location in every timepoint. These landmarks could be used to compute registrations between arbitrary time points.

This solutions could be implemented in two parts and be really flexible:
1. one plugin that allow to select spot and creates a landmark from the selected spots. Simply group the spots into timepoints and average their position for each timepoint. Create a new spot on the average position, connect all the spots. ([this method](https://github.com/mastodon-sc/mastodon-tomancak/blob/fcd2d45880cd9be67a3e5ea4276625eee715a0af/src/main/java/org/mastodon/mamut/tomancak/sort_tree/SortTreeUtils.java#L146) could be used to implement this)
2. Let the lineage registration plugin use a landmarks tagset to register two embryos over time.

## Actual Solution
- [x] There's a new plugin in the Mastodon menu "Plugins > Tree Management" > "Add center spot". It adds a new spot at the center (mean location) of the selected spots. If spots are selected in multiple timepoints, then there will be a new spot for each timepoint at the mean location. The spots are connected with edges as usual.
- [x] There is now a new combo box in the lineage registration dialog. It has three options:
  - **"fixed spatial registration based on root cells"**. The algorithm works exactly as before if this item is selected. (for example "tag unmatched and flipped cells", will create the same tags as before)
  - **"dynamic spatial registration based on root cells and their descendants"**. This will give better results for rotating embryos. The algorithm will compute an individual coordinate transform for the registration of each cell division. The root cells and their descendants are used as landmarks to compute the coordinate transform. 
  - **"dynamic spatial registration based on 'landmarks' tag set."** This also allows to get better results for rotating embryos. But the user has full control over the landmarks that are used. Both datasets need to have a tag set named "landmarks" and the tags in both datasets need to be the same. The spots that are tagged with those tags are used as landmarks to derive a coordinate transform. (A tip for testings. Simply click the button "color paired lineages" and rename the created tag sets "lineages" -> "landmarks". With this tagset the algorithm will give the same result as "dynamic spatial registraion base on roots..."

### PR DESCRIPTION
**Pull request by maarzt**
Closes #22

### CODE REVIEW

**Review by stefanhahmann — COMMENTED** [review]
This PR modifies the method how the lineage registration is done and adds new methods.

- It is important to mention that this PR is not 100% in line with the issue description (https://github.com/mastodon-sc/mastodon-tomancak/issues/22). The "fixed" option is modified. It seems to improve the registration, but is not working exactly as before, as the issue description claims, but this can be solved by providing some documentation about the three registration methods.

- It is hard to judge, if the dynamic registration really improves the registration. It seems so, but one can also find examples, where it worsens the registration.

- Since the "Color paired lineages" function is now the only function that is independent from the chosen registration, it may be moved to different dialog. Users otherwise may wonder why this function always leads to the same result, no matter, which registration method is active.

- In TrackScheme, there is a bug, when adding a new center Spot. TrackScheme needs to be refreshed with `Ctrl` + `Del` after a new center spot is added, before the user can see it in TrackScheme.

- Regarding registration with the "landmarks" tag set: the documentation should inform users that in case of multiple "landmarks" tag set, the first one is chosen by the algorithm.

- Please consider renaming "spa*c*ial" to "spa*t*ial". Both versions of the word exist, but the version with 't' seems to be the more common one

**Comment by maarzt** [comment]
@stefanhahmann Thank you for testing the code and sharing your opinions. I agree with your concerns. Some problems are fixed now. Here are my replies to the individual points. I think the PR is good enough to be merged now.

> - It is important to mention that this PR is not 100% in line with the issue description (https://github.com/mastodon-sc/mastodon-tomancak/issues/22). The "fixed" option is modified. It seems to improve the registration, but is not working exactly as before, as the issue description claims, but this can be solved by providing some documentation about the three registration methods.

That's true. There is a small different between the old behavior and the "fixed spacial registration" option. The old behavior used spot coordinates on timepoint 0 for registration. The new behavior uses coordinates of the root cells just before cell division.

> - It is hard to judge, if the dynamic registration really improves the registration. It seems so, but one can also find examples, where it worsens the registration.

I agree that it is hard to judge. But there are reasons to believe that the dynamic registration works better:
* Some embryos rotate. For them the fixed registration is expected to be worse. The unit test [DynamicLandmarkRegistrationTest](https://github.com/mastodon-sc/mastodon-tomancak/blob/dynamic-spacial-registration/src/test/java/org/mastodon/mamut/tomancak/lineage_registration/spatial_registration/DynamicLandmarkRegistrationTest.java) demonstrates that the "dynamic registration" works on a artificial rotating dataset where "fixed registration" would fail.
* I did a quantitative comparison by measuring angles between cell division directions between the two embryos. I did this across all dataset that I had available. "dynamic registration" led to slightly smaller angles on average. (Smaller angles is better). The difference is small, and for some datasets the angles actually worsened. I unfortunately lost the code for this comparison :facepalm: 

> - Since the "Color paired lineages" function is now the only function that is independent from the chosen registration, it may be moved to different dialog. Users otherwise may wonder why this function always leads to the same result, no matter, which registration method is active.

I agree with your concerns. But a separate dialog, where you select dataset and timepoints, is not worth the struggle. Lets instead write documentation that makes things clear.

> - In TrackScheme, there is a bug, when adding a new center Spot. TrackScheme needs to be refreshed with `Ctrl` + `Del` after a new center spot is added, before the user can see it in TrackScheme.

Fixed in [#e1dce454c1cf3eba6ebfb4888dbb38f4cad6b4ef](https://github.com/mastodon-sc/mastodon-tomancak/pull/28/commits/e1dce454c1cf3eba6ebfb4888dbb38f4cad6b4ef)

> - Regarding registration with the "landmarks" tag set: the documentation should inform users that in case of multiple "landmarks" tag set, the first one is chosen by the algorithm.

I agree.

> - Please consider renaming "spa*c*ial" to "spa*t*ial". Both versions of the word exist, but the version with 't' seems to be the more common one

done

**Review by stefanhahmann — CHANGES_REQUESTED** [review]
There is a comment in a related PR that does not seem to have been addressed:
https://github.com/mastodon-sc/mastodon-tomancak/pull/27#discussion_r1193658174

The CI seems to fail on DynamicLandmarkRegistrationTest.testLineageRegistrationAlgorithm()
It is not clear to me, if this is related to this PR or perhaps to https://github.com/mastodon-sc/mastodon/issues/225
Could you please check?

File: src/main/java/org/mastodon/mamut/tomancak/lineage_registration/LineageRegistrationUtils.java [hunk 1]
```diff
@@ -35,9 +35,8 @@
 	 * The sorting is based on the result of the
```
[the remark this answers is not in the document]
Comment (maarzt) [reply src/main/java/org/mastodon/mamut/tomancak/lineage_registration/LineageRegistrationUtils.java:33]: Thank you for spotting this it's fixed now.

File: src/main/java/org/mastodon/mamut/tomancak/lineage_registration/RefCollectionUtils.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
+package org.mastodon.mamut.tomancak.lineage_registration;
+
+import java.util.function.Predicate;
+import java.util.function.UnaryOperator;
+
+import org.mastodon.RefPool;
+import org.mastodon.collection.RefCollection;
+import org.mastodon.collection.RefCollections;
+import org.mastodon.collection.RefRefMap;
```
[the remark this answers is not in the document]
Comment (maarzt) [reply src/main/java/org/mastodon/mamut/tomancak/lineage_registration/RefCollectionUtils.java]: Thank you for spotting this. It's fixed now.

**Pull request merged**